text/3692-feature-unification.md

```diff
@@ -0,0 +1,193 @@
+
+We'll add two new modes to feature unification:
+
+**Unify features across the workspace, independent of the selected packages**
+
+This would be built-in support for "cargo-workspace-hack".
+
+This would require effectively changing from
+1. Resolve dependencies
+2. Filter dependencies down for current target and selected packages
+3. Resolve features
+
+To
+1. Resolve dependencies
+2. Filter dependencies down for current target
+3. Resolve features
+4. Filter for selected packages
+
+**Features will be evaluated for each package in isolation**
+
+This will require building duplicate copies of build units when there are disjoint sets of features.
+
+For example purposes., this could be implemented as either
+- Loop over the packages, resolving, and then run a build plan for that package
+- Resolve for each package and generate everything into the same build plan
+
+This is not prescriptive of the implementation but to illustrate what the feature does.
+The initial implementation may be sub-optimal.
+Likely, the implementation could be improved over time.
+
+**Note:** these features do not need to be stabilized together.
+
+##### `resolver.feature-unification`
+
+*(update to [Configuration](https://doc.rust-lang.org/cargo/reference/config.html))*
+
+* Type: string
+* Default: "selected"
+* Environment: `CARGO_RESOLVER_FEATURE_UNIFICATION`
+
+Specify which packages participate in [feature unification](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification).
+
+* `selected`: merge dependency features from all package specified for the current build
+* `workspace`: merge dependency features across all workspace members, regardless of which packages are specified for the current build
+* `package`: dependency features are only enabled for each package, preferring duplicate builds of dependencies to when different feature sets are selected
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This increases entropy within Cargo and the universe at large.
+
+If someone enables mutually exclusive features in different packages, then `workspace` unification will fail.
+Officially, features are supposed to be additive, making mutually exclusive features officially unsupported.
+Instead, effort should be put towards [official mutually exclusive globals](https://internals.rust-lang.org/t/pre-rfc-mutually-excusive-global-features/19618).
+
+Some features cannot be enabled in some packages, like a `no_std` package not wanting `std` features.
+These workspaces will not be able to use `workspace` unification.
+For now, unifying for the `"workspace"` is primarily targeted at single-application workspaces.
+The other config fields can always be used instead.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This is done in the config instead of the manifest:
+- As this can change from run-to-run, this covers more use cases
+- As this fits easily into the `resolver` table. there is less design work
```

Comment:
```suggestion
- As this can change from run to run, this covers more use cases.
- As this fits easily into the `resolver` table, there is less design work.
```